src/serac/physics/contact/contact_data.cpp

```diff
@@ -0,0 +1,367 @@
+// Copyright (c) 2019-2023, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "serac/physics/contact/contact_data.hpp"
+
+#include "axom/slic.hpp"
+
+#ifdef SERAC_USE_TRIBOL
+#include "tribol/interface/tribol.hpp"
+#include "tribol/interface/mfem_tribol.hpp"
+
+namespace serac {
+
+ContactData::ContactData(const mfem::ParMesh& mesh)
+    : mesh_{mesh},
+      reference_nodes_{dynamic_cast<const mfem::ParGridFunction*>(mesh.GetNodes())},
+      current_coords_{*reference_nodes_},
+      have_lagrange_multipliers_{false},
+      num_pressure_dofs_{0}
+{
+  tribol::initialize(mesh_.SpaceDimension(), mesh_.GetComm());
+}
+
+ContactData::~ContactData() { tribol::finalize(); }
+
+void ContactData::addContactInteraction(int interaction_id, const std::set<int>& bdry_attr_surf1,
+                                        const std::set<int>& bdry_attr_surf2, ContactOptions contact_opts)
+{
+  interactions_.emplace_back(interaction_id, mesh_, bdry_attr_surf1, bdry_attr_surf2, current_coords_, contact_opts);
+  if (contact_opts.enforcement == ContactEnforcement::LagrangeMultiplier) {
+    have_lagrange_multipliers_ = true;
+    num_pressure_dofs_ += interactions_.back().numPressureDofs();
+  }
+}
+
+void ContactData::update(int cycle, double time, double& dt)
+{
+  // This updates the redecomposed surface mesh based on the current displacement, then transfers field quantities to
+  // the updated mesh.
+  tribol::updateMfemParallelDecomposition();
+  // This function computes forces, gaps, and Jacobian contributions based on the current field quantities. Note the
+  // fields (with the exception of pressure) are stored on the redecomposed surface mesh until transferred by calling
+  // forces(), mergedGaps(), etc.
+  tribol::update(cycle, time, dt);
+}
+
+FiniteElementDual ContactData::forces() const
+{
+  FiniteElementDual f(*reference_nodes_->ParFESpace(), "contact force");
+  for (const auto& interaction : interactions_) {
```

Comment:
This this pulling forces from the parallel contact decomp?  And if so, 1 MPI round per interaction?

Reply:
Yes and yes